### DISCUSSION TITLE
Reorganize the infra repo with added support to GCE

### DISCUSSION THREAD

**Discussion by merlinran** [testground/testground#1238] [Ideas]
First of all, thank you for making this powerful yet super easy-to-use test tool! We at [Textile](https://textile.io/) have been thinking about using it for months.

This week I've been migrating the https://github.com/testground/infra scripts to support GCE. So far I've been able to launch k8s, add block storage and file storage via GCE features. Now I'm thinking about ways to reorganize the code to support both GCE and AWS (and allow adding others in the future). The plan is:
1. move provider specific config to their own subdirectories, i.e., `k8s/aws/efs/...`, `k8s/gce/filestore/...`, etc.
2. extract provider specific shell code to functions.
3. rename all envvars to be provider neutral, then add a new one `CLOUD_PROVIDER`.
4. merge the shell scripts back to a single one with subcommands, like `infra install all`, `infra install blockstore`, `infra delete filestore`, etc.

So here are the questions to the testground team before I start working toward a PR:
1. Are you guys okay with adding support to GCE and others in the future?
2. Any opinions and/or recommendations on the proposed code refactoring?

Oh one more thing - to support new provider, are there any other additions / modifications to the testground daemon code needed other than https://github.com/testground/testground/tree/master/pkg/aws ?

**Comment by nonsense** [comment]
Hey @merlinran !

Thank you for your detailed explanation!

Support for GCE would be very welcome! What you are proposing as a plan, sounds good to me!

> Oh one more thing - to support new provider, are there any other additions / modifications to the testground daemon code needed other than https://github.com/testground/testground/tree/master/pkg/aws ?

There will be need for slight refactor with respect to pushing Docker images, as well as the `cluster:k8s` runner. It is not just the `./pkg/aws` that would need to be changed.

Let me know if you have any other questions.
